extensions/2.0/Vendor/EXT_meshopt_compression/README.md

```diff
@@ -0,0 +1,480 @@
+
+	x += copysign(t, x);
+	y += copysign(t, y);
+
+	// renormalize (x, y, z)
+	float32_t l = sqrt(x * x + y * y + z * z);
+
+	x /= l;
+	y /= l;
+	z /= l;
+
+	output[0] = round(x * INTN_MAX);
+	output[1] = round(y * INTN_MAX);
+	output[2] = round(z * INTN_MAX);
+	output[3] = input[3];
+}
+```
+
+## Filter 2: quaternion
+
+Quaternion filter allows to encode unit length quaternions using normalized 16-bit integers for all components, but allows control over the precision used for the components and provides better quality compared to naively encoding each component one by one.
+
+The input to the filter is three quaternion components, excluding the component with the largest magnitude, encoded as signed normalized K-bit integers (4 <= K <= 16), and an index of the largest component that is omitted in the encoding. The largest component is assumed to always be positive (which is possible due to quaternion double-cover). To allow per-element control over K, the last input element explicitly encodes 1.0 as a K-bit integer, except for the least significant 2 bits that store the index of the maximum component.
+
+The output of the filter is four decoded quaternion components, stored as 16-bit normalized integers.
+
+After eliminating the maximum component, the maximum magnitude of the remaining components is 1/sqrt(2). Because of this the input components store the original component value scaled by sqrt(2.0) to increase precision.
+
+```
+void decode(int16_t input[4], int16_t output[4]) {
+	float32_t range = 1.0 / sqrt(2.0);
+
+	// input[3] encodes a K-bit representation of 1.0 except for bottom two bits
+	float32_t one = input[3] | 3;
+
+	float32_t x = input[0] / one * range;
+	float32_t y = input[1] / one * range;
+	float32_t z = input[2] / one * range;
+
+	float32_t w = sqrt(max(0.0, 1.0 - x * x - y * y - z * z));
+
+	int maxcomp = input[3] & 3;
+
+	// maxcomp specifies a cyclic rotation of the quaternion components
+	output[(maxcomp + 1) % 4] = round(x * 32767.0);
+	output[(maxcomp + 2) % 4] = round(y * 32767.0);
+	output[(maxcomp + 3) % 4] = round(z * 32767.0);
+	output[(maxcomp + 0) % 4] = round(w * 32767.0);
+}
+```
+
+## Filter 3: exponential
+
+Exponential filter allows to encode floating point values with a range close to the full range of a 32-bit floating point value, but allows more control over the exponent/mantissa to trade quality for precision, and has a bit structure that is more optimally aligned to the byte boundary.
+
+The input to the filter is a sequence of 32-bit little endian integers, with the most significant 8 bits specifying a (signed) exponent value, and the remaining 24 bits specifying a (signed) mantissa value. The integers are stored in two-complement format.
+
+The result of the filter is 2^e * m:
+
+```
+float32_t decode(int32_t input) {
+	int32_t e = input >> 24;
+	int32_t m = (input << 8) >> 8;
+	return pow(2.0, e) * m;
+}
+```
+
+The valid range of `e` is [-100, +100], which facilitates performant implementations.

```

Comment:
What should decoders do with values out of this range?

Reply:
The behavior is unspecified (this range limit exists to make it possible for decoders to simply shift the exponent into the 32-bit floating point range without having to handle overflow conditions that require renormalizing the mantissa), left a comment to that effect + noted that encoders are expected to clamp.